Adding in support for SDL2_Mixer

jwoertink opened the pull request:
Ok, so this is an initial PR with the very first basic sound working. I have a lot of the code for the rest of it in here, but I get random errors like `Invalid memory access` or `Undefined symbols for architecture x86_64:`. I'm hoping maybe you can spot what I'm doing wrong getting those errors.

But, as for the working example, I have added [sample 21](http://lazyfoo.net/tutorials/SDL/21_sound_effects_and_music/index.php). It will require getting the `prompt.png`, and sound files `low.wav`, `medium.wav`, `high.wav`, and `beat.wav`. You can run the sample with `crystal samples/21_sound_and_effects.cr`. Once booted, pressing the 9 key will start the music, then pressing 9 again will pause it.

ysbaddaden commented:
As outlined on Gitter by Matyn Jago, `SDL_Load_WAV` is a C macro, not a method:
https://gist.github.com/mjago/188964f13a39813afecea75a43372f67

Anyway, you should always use the `_RW` method variants; RWops uses Crystal IO and thus trigger the event loop as needed.

If some other symbols are missing, they probably are C macros (look at the C headers).

jwoertink commented:
Ok, made a few updates here. Interesting thing is, with the `load_music`, when I used the block form:
```crystal
def self.load_music(filename, type : Type? = nil)
  SDL::RWops.open(filename, "rb") do |rwops|
    if type
      audio = LibMixer.load_mus_type_rw(rwops, type.to_s, 1)
      raise SDL::Error.new("Mix_LoadMUSType_RW") unless audio
    else
      audio = LibMixer.load_mus_rw(rwops, 1)
      raise SDL::Error.new("Mix_LoadMUS_RW") unless audio
    end
    audio
  end
end
```

then I get `Invalid memory access (signal 10) at address 0x7fffd01b7780`. Switching this to just use:

```crystal
rwops = LibSDL.rw_from_file(filename, "rb")              
if type                                                  
  audio = LibMixer.load_mus_type_rw(rwops, type.to_s, 1) 
  raise SDL::Error.new("Mix_LoadMUSType_RW") unless audio
else                                                     
  audio = LibMixer.load_mus_rw(rwops, 1)                 
  raise SDL::Error.new("Mix_LoadMUS_RW") unless audio    
end                                                      
audio                                                    
```

this plays fine. However, the `load_wav` still breaks with `ld: symbol(s) not found for architecture x86_64`.

jwoertink commented:
Ok, I figured out what I was doing wrong. All the sample sounds are playing for me now!

ysbaddaden commented on src/mixer.cr:
```diff
@@ -40,17 +40,15 @@ module SDL
+        audio = LibMixer.load_mus_rw(rwops, 1)
+        raise SDL::Error.new("Mix_LoadMUS_RW") unless audio
+      end
+      audio
```
Any reason to avoid `SDL::RWops`?

jwoertink replied:
I was using the block form `SDL_RWops.open` and getting an Invalid memory access error. Using it like this works for me with no errors.

ysbaddaden replied:
I assume the file is kept open by SDL, then streamed from, which would explain why it segfaults when the block form is used (the RWops is freed).

The problem is that it's still open on the stack, and there is no reference kept to it, so it will probably crash later, when the GC collects memory.

I think we need a wrapping class, here, that will keep both the rwops and audio references (and release them in a finalize def).

jwoertink replied:
Hmm, so then it becomes an instance of this wrapper class, and those are assigned to ivars? Then the API goes from `SDL::Mixer.load_music(filename)` to something like `SDL::Mixer::Sound.new(filename)`?

ysbaddaden replied:
Or `sample = Mixer.load(filename)` maybe? as a factory for `Mixer::Sample` for example.

Then maybe introduce a nice API to play/pause/seek a sample, like `sample.play`, `sample.pause`, ... or something along those lines.

jwoertink replied:
I like that. The only thing I can think of that we need to know weather a mus or wav is being loaded. Unless we say that the `Sample` class is specifically for loading a mus because of all the additional functions like stop, play, pause, etc... Then we would just leave the wav stuff as is.

ysbaddaden replied:
I guess I don't know enough of the API, yet. There are mus, wav, channels, ...

jwoertink replied:
Well, the 2 main things are `mus` and `wav`. `mus` is something longer like a song. `wav` is something shorter like a sound. `channels` just refers to like "left" or "right" channel for stereo or mono, etc...

The main thing is deciding if we use something like `SDL::Sample.new(some_filename)` are we making an object to surround just a song, or do have an optional overload like `SDL::Sample.new(some_filename, :wav)` to allow for either one? I'm not too familiar with how the GC works, so from what you can tell, does the mus portion fall in to the same issue of possibly being GC'd? My guess is not in which case we could just wrap `mus` for now.

ysbaddaden replied:
If Mixer does such a difference in types, we should do it too, with different types, because accessible methods will be very different, aren't they?

jwoertink replied:
For the most part, the interface between the two are the same. The only major difference is that you can't "pause" or "resume" a `wav`. But honestly, the API for SDL2_Mixer is really bad. It's inconsistent in how things are implemented. Like a `wav` is referred to as `chunk` in half the functions, but they are the same thing. We could always do 2 separate wrapper classes, but then the top level factory method would need to be something like `Mixer.load_sample` and `Mixer.load_song` or whatever you wanted to call them..

Maybe we can ask @mjago for some additional input.

mjago replied:
The obvious difference between MUS and WAV is long-playing (streamed) bg vs short, pre-loaded / freeable effects. I think the difference should remain clear. My interest was to get cover art into an audio player / streamer but not really suitable tbh. Mp3 MUS needs stripping of tags to play successfully for instance. For games programming succeeds in abstracting away drivers etc. (disclaimer: have no experience in playing or writing games).

ysbaddaden replied:
Thanks for the clarification:

- Music is for background music: plays along, only one at a time, loops forever;
- WAV are sound effects: as many as needed can be mixed together through channels.

We definitely need to keep the difference, but we should introduce a better naming (`Music` or `Song` vs `Sample`).

I'm not sure which API would be the best, though. I think I'll prefer `Mixer#play(music)` (is played globally), but having `Sample#play(channel = -1)` (can play concurrently).

ysbaddaden replied:
Or maybe introduce `Channel` too, then:

```crystal
boom = Mixer.load_sample("data/boom.ogg")

channel = Channel.new(1)
channel.volume = 10
channel.play(boom)

# later
channel.play # => boom
```

jwoertink replied:
Ok, that sounds good. Is this something we could maybe just merge this PR in to master for now, then create a separate PR to sort of clean, refactor and abstract out the factory stuff? Or do you want these added in to this PR?

mjago replied:
Well I had some time this weekend and implemented Music, Channel, and Sample. Any feedback would be welcome. The changes can be seen [here](https://github.com/jwoertink/sdl.cr/compare/audio...mjago:audio)

jwoertink commented:
A few additions from @mjago for volume and fading.

jwoertink commented:
@ysbaddaden Not sure if you saw, but @mjago added some new stuff [here](https://github.com/jwoertink/sdl.cr/compare/audio...mjago:audio). Since it's sort of a big change to things, let me know if you think I should merge it in.

ysbaddaden commented:
Quick review, I don't have much type for a thorough review:

- Don't use aliases:
  - drop the `LMix` private alias, use the actual `LibMixer`;
  - drop the `Mixer` alias in `samples/` use `SDL::Mixer` instead;

Note that if using `SDL::MIX`, `SDL::IMG` or `SDL::TTF` it too bothersome or noisy (I agree), we should consider dropping the `SDL` namespace altogether, not using aliases in samples, which proves they're merely a bother.

- Namings:
  - `LibMixer` should be `LibMIX`;
  - `SDL::Mixer` should be `SDL::MIX`.

- Drop `free` methods, add class finalizers instead:
  - `Music#free` => `Mixer#finalize`
  - `Sample#free` => `Sample#finalize`
  - ...

- Never take/expose C pointers:
  - Rewrite `Mixer.query_spec` so it initializes stack based values (in the function), and returns a tuple or struct.

- Drop/avoid blocking calls:
  - drop `Sample#quick_load` (blocking, confusing)
  - `Sample#load(filename)` must use `SDL::RWops` (evented)
  - `Music#load(filename)` must use `SDL::RWops` (evented)

jwoertink commented:
Ok, latest update includes most of these changes.
* No more private alias
* Using full namespaces
* Renamed `LibMixer` => `LibMIX`, `SDL::Mixer` => `SDL::MIX`
* Moved method `Music#free` => `Music#finalize`, `Sample#free` => `Sample#finalize`
* Dropped method `Sample#quick_load`.

The sample works with this update, but the only things left are implementing the `MIX.query_spec` method so it returns a tuple and doesn't expose C pointers. This method isn't required for the sample to run (and I'm not really sure what it does anyway). Lastly, the `load` methods will just need to be refactored.

ysbaddaden commented on samples/21_sound_and_effects.cr:
```diff
@@ -22,35 +26,38 @@ activekey = [] of LibSDL::Keycode
   when SDL::Event::Quit
-    #SDL::Mixer.unload_music(music)
+    music.stop
+    SDL::MIX.close
```
Do we have to close the mixer? The `at_exit` above will already call `MIX.quit`.

jwoertink replied:
oh, good catch. No, we don't have to call it there.

jwoertink replied:
Well, actually, they do different things. `quit` closes out the decoders, and `close` closes out the music player. However, the original sample I took this from only calls `Mix_Quit()`. I'm assuming you would use the `MIX.close` when you want to change songs? I'll remove that line since it's not necessary for this sample.

mjago commented:
Thanks @jwoertink 👍  I'm very busy with work related stuff

jwoertink commented:
Alright, just added in the query_spec method refactored without exposing pointers. I tried to refactor the load methods, but using block form of RWops, it throws exceptions. I think they should be fine with how they are now since it still uses rwops internally. Just not in a block form. 

Let me know if you see anything else missing.

jwoertink commented:
@ysbaddaden Have you had a chance to look over these last changes? 😄  I'm itchin to get these changes merged so I can start playing with it lol

ysbaddaden commented on README.md:
```diff
@@ -7,3 +7,42 @@ Most of the bindings have been implemented, though most haven't been verified,
 found at <http://lazyfoo.net/tutorials/SDL/index.php>. You are welcome to port
 more samples, and the necessary corrections!
+
+## Requirements
```
Please drop the library requirements. Only `libSDL2` by-the-way, other libraries are optional.

Also drop the homebrew commands; SDL2 is multiplatform and we shouldn't have to tell how to install the libraries. One may choose the solution that best fits them (distrib packages, manual compilations, ...).

ysbaddaden commented on samples/README.md:
```diff
@@ -2,5 +2,5 @@
 
-BPM and PNG images aren't included, to avoid bloating the repository. Download
-and extract them as see fit.
+ data files aren't included, to avoid bloating the repository. Download
```
It should be `Data` not `[space]data`

jwoertink commented:
@ysbaddaden Ok, I think I rebased it correctly 😂 haha. I made the changes to those README files. Let me know if you see anything else!

ysbaddaden commented on src/mix.cr:
```diff
@@ -0,0 +1,255 @@
+      private getter music : Pointer(LibMIX::Music) | Nil
+
+      def initialize(filename = nil)
+        @music = filename ? load(filename) : nil
```
Same here: `MIX::Music#initialize` should require a filename and always load a music file; `@music` doesn't have to be a nilable, and `#finalize` doesn't have to set it to nil (GC is collecting the memory anyway).

Unless there is something I didn't understand about SDL_Mixer... like we may only play a single BGM music at once, in which case we should have an argless `MIX::Music#initialize` (i.e. remove the method) have `#load` load or replace the current music (which may require to free any previous `@music`?) then verify that `@music` is set before calling `free_music` in `#finalize`.

I even wonder if it should be a class at all, or just a module with class methods? There is no state, except the currently loaded `SDL::RWops` and `SDL::MIX::Music` which could be class variables.

ysbaddaden commented on src/mix.cr:
```diff
@@ -0,0 +1,255 @@
+      end
+
+      def load(filename, type : Type? = nil)
+        rwops = LibSDL.rw_from_file(filename, "rb")
```
Same as above: it should use `SDL::RWops.new(filename, "rb")` and keep a reference to it, so GC won't collect prematurely.

ysbaddaden commented on src/mix.cr:
```diff
@@ -0,0 +1,255 @@
+  end
+
+  module MIX
+    class Channel
```
Maybe we could have delegations to an `ANY` channel (-1)? instead of duplicating methods as class methods with an hardcoded `-1` channel?

```crystal
module Channel
  ANY = Channel.new(-1)
  
  delegate play, resume, expire, fade_in, fade_out, volume, volume=, ..., to: ANY
end
```

ysbaddaden commented on src/mix.cr:
```diff
@@ -0,0 +1,255 @@
+      property id
+
+      def initialize(@id = 1)
+        @@count ||= 8 unless @@count
```
`#initialize` shouldn't initialize class variables:

```crystal
class Channel
  @@count = 0
  @@reserved = 0
end
```

ysbaddaden replied:
There may be something I don't know happening here, thought. Like by default 8 channels are opened? So maybe `@@count` should be set to 8 by default instead of 0?

ysbaddaden commented on src/mix.cr:
```diff
@@ -0,0 +1,255 @@
+      end
+
+      def self.play(smpl, repeats = 0)
+        LibMIX.play_channel(-1, smpl.sample, repeats)
```
`MIX::Sample` shouldn't expose `#sample` but declare `#to_unsafe` so we can just pass an instance of `Sample` to C bindings, and Crystal will call `#to_unsafe` automatically. This way we could use `sample` as a variable name, which is nicer than `smpl`.

ysbaddaden commented on src/mix.cr:
```diff
@@ -0,0 +1,255 @@
+
+      private def load_music(rwops)
+        audio = LibMIX.load_mus_rw(rwops, 1)
+        raise SDL::Error.new("Mix_LoadMUSType_RW") unless audio
```
Typo: you interverted the method names `Mix_LoadMUSType_RW` with `Mix_LoadMUS_RW` below

ysbaddaden commented on src/mix.cr:
```diff
@@ -0,0 +1,255 @@
+
+  module MIX
+    class Sample
+      protected getter sample
```
You may drop the getter, use `@sample` and expose it as `#to_unsafe` only, so you may pass `self` to C bindings. This would follow what the design of other SDL wrapper classes.

jwoertink commented:
I'm making these changes, but running in to some issues. For the `MIX::Sample`, if I move that load in to the initializer, then use the `@rwops = SDL::RWops.new(filename, "rb")` I get the error:

```
Mix_LoadWAV_RW: Unrecognized file type (not WAVE) (SDL::Error)
```
Using that exact same code, but from the load method, I get the error:

```
LibMIX#load_wav_rw' must be Pointer(LibSDL::RWops), not (SDL::RWops | Nil)
```

I'll keep trying to work through that, but if you have some insight, that would be great.
Also, do you want me to rename `MIX` -> `Mix`? as well as `LibMIX` -> `LibMix`?

jwoertink commented:
@ysbaddaden ok, so I've made most those changes. I'm not sure if you wanted `MIX` to be renamed to `Mix`, but that's an easy fix. As for the RWops, It just doesn't work with Sample, and with Music it has weird side effects. I feel like those changes could be made after this branch is merged. That way this can get out there and start getting used, and then optimizations can be made after. Let me know what your thoughts are. Thanks :)

ysbaddaden commented on samples/21_sound_and_effects.cr:
```diff
@@ -0,0 +1,62 @@
+require "../src/mix"
+
+SDL.init(SDL::Init::VIDEO | SDL::Init::AUDIO); at_exit { SDL.quit }
+SDL::MIX.init(SDL::MIX::Init::MP3); at_exit { SDL::MIX.quit }
```
I can't init with MP3 on Ubuntu with the system packages, because SDL_mixer is compiled without MP3 support. Using `FLAC` instead works. I'm surprised there is no `WAV` type.

ysbaddaden commented on src/mix.cr:
```diff
@@ -0,0 +1,260 @@
+
+    # This is required to initial SDL_Mixer. It must be called before using any other function, but AFTER SDL has been initialized.
+    def self.open(freq = 44100, format = 0, channels = 2, sample_size = 2048)
+      format = format == 0 ? LibMIX::MIX_DEFAULT_FORMAT : format.to_u16
```
Maybe: 
```crystal
def open(freq = 44100, format = LibMix::MIX_DEFAULT_FORMAT, ...)
end
```

ysbaddaden commented:
Manually merged as d40de3177809ac2f71b5980632b422565eb0a3b4. Thanks for hard work @jwoertink!

jwoertink commented:
YAY! 🎉
